Clarify how ranking Qualifications are selected.

timreyn opened the pull request:
The current spec does not distinguish between "Top N average" and "Top N single". Some competitions use average while others use single for events like 333fm.

Related to thewca/worldcubeassociation.org#2051.

gregorbg commented:
Fine by me! I like how this allows to potentially introduce even more `method`s in the future.
What are your thoughts on this @jonatanklosko ?

timreyn commented:
One other method I could include is any_result (i.e. "Any WCA Result"). This would not have an associated "type". What do you think?

This technically doesn't need to be a separate method, since today it can be implemented as `single < 2^31-1`, but it's probably clearer to have it as a separate method.

gregorbg commented:
Having `any_result` sounds great, as I have definitely seen that in practice as well!
However, why would it not have an associated `type`? Shouldn't it be "legal" to specifically require having an average result, i.e. in 6x6 or 7x7?

Also, implementing it as `single < 2^31-1` is technically possible but then some (external) programs that use WCIF for their own purposes might mis-interpet it as "single below 8.16 months" (lol). This is not only meant as a joke but also as a legitimate concern about correctly displaying qualification results when you don't know that there's this 32-bit int max "magic number" to signify "any result"

timreyn commented:
Fair points (and yes, that's the kind of reason which is why I think it's clearer to have a separate method). Added!

gregorbg commented on specification.md:
```diff
@@ -395,27 +395,38 @@ See paragraph 5.1 of [WCA Competition Requirements Policy](https://www.worldcube
+{
+  "whenDate": "2020-04-25",
+  "type": "single",
+  "method: "any_result",
```
Missed a quotation mark

timreyn replied:
Ah, thank you! Fixed.

jonatanklosko commented on specification.md:
```diff
@@ -395,27 +395,38 @@ See paragraph 5.1 of [WCA Competition Requirements Policy](https://www.worldcube
-| `level` | [`AttemptResult`](#attemptresult)\|[`Ranking`](#ranking) | The parameter of the qualification condition of the given type. |
+| `type` | `"single"\|"average"` | The type of result the requirement refers to. |
+| `method` | `"result"\|"ranking"\|"any_result"` | The method for determining which competitors qualify. |
+| `level` | [`AttemptResult`](#attemptresult)\|[`Ranking`](#ranking)\|null | The parameter of the qualification condition of the given type. Not used with `any_result`. |
```
```suggestion
| `level` | [`AttemptResult`](#attemptresult)\|[`Ranking`](#ranking)\|`null` | The parameter of the qualification condition of the given type. Not used with `any_result`. |
```

jonatanklosko commented on specification.md:
```diff
@@ -395,27 +395,38 @@ See paragraph 5.1 of [WCA Competition Requirements Policy](https://www.worldcube
-| `type` | `"single"\|"average"\|"ranking"` | The type of result the requirement refers to. |
-| `level` | [`AttemptResult`](#attemptresult)\|[`Ranking`](#ranking) | The parameter of the qualification condition of the given type. |
+| `type` | `"single"\|"average"` | The type of result the requirement refers to. |
+| `method` | `"result"\|"ranking"\|"any_result"` | The method for determining which competitors qualify. |
```
To stay consistent with `AdvancementCondition` let's use `"attemptResult"` instead of `"result"`. Also, the spec generally uses camel case, so `any_result -> anyResult`.

Looking at `AdvancementCondition`, I'm also wondering: `method -> type`, `type -> resultType/variant`, thoughts?

jonatanklosko reviewed (commented):
Sounds good to me! Just a couple comments :)

timreyn committed:
Clarify how ranking Qualifications are selected.

The current spec does not distinguish between "Top N average" and "Top N single". Some competitions use average while others use single for events like 333fm.

In addition, add an "anyResult" qualification type.

Related to thewca/worldcubeassociation.org#2051.

jonatanklosko commented on specification.md:
```diff
@@ -412,10 +414,19 @@ See paragraph 5.1 of [WCA Competition Requirements Policy](https://www.worldcube
+{
+  "whenDate": "2020-04-25",
+  "type": "anyResult",
+  "resultType": "single",
```
```suggestion
  "resultType": "single",
  "level": null
```

timreyn commented:
Thanks for the review @jonatanklosko !

jonatanklosko commented on specification.md:
```diff
@@ -395,15 +395,17 @@ See paragraph 5.1 of [WCA Competition Requirements Policy](https://www.worldcube
 | `whenDate` | [`Date`](#date) | The date by which the qualification requirement must be satisfied.  If a result is set in a multiple-day competition which begins by this date, that is considered to have been set by this date. |
-| `type` | `"single"\|"average"\|"ranking"` | The type of result the requirement refers to. |
-| `level` | [`AttemptResult`](#attemptresult)\|[`Ranking`](#ranking) | The parameter of the qualification condition of the given type. |
+| `type` | `"attemptResult"\|"ranking"\|"anyResult"` | The type of Qualification standard. |
```
```suggestion
| `type` | `"attemptResult"\|"ranking"\|"anyResult"` | The type of qualification. Either of ranking (top N competitors), attempt result (competitors with result better than Y - either single or average) or any result (competitors with any successful result of the given type). |
```

jonatanklosko committed:
Update specification.md
